In [1]:
import pandas as pd

url = "https://raw.githubusercontent.com/4GeeksAcademy/naive-bayes-project-tutorial/main/playstore_reviews.csv"
df = pd.read_csv(url)
df

package_name                                             review  \
0     com.facebook.katana   privacy at least put some option appear offli...   
1     com.facebook.katana   messenger issues ever since the last update, ...   
2     com.facebook.katana   profile any time my wife or anybody has more ...   
3     com.facebook.katana   the new features suck for those of us who don...   
4     com.facebook.katana   forced reload on uploading pic on replying co...   
..                    ...                                                ...   
886  com.rovio.angrybirds   loved it i loooooooooooooovvved it because it...   
887  com.rovio.angrybirds   all time legendary game the birthday party le...   
888  com.rovio.angrybirds   ads are way to heavy listen to the bad review...   
889  com.rovio.angrybirds   fun works perfectly well. ads aren't as annoy...   
890  com.rovio.angrybirds   they're everywhere i see angry birds everywhe...   

     polarity  
0           0  
1           0  
2           0  
3           0  
4           0  
..        ...  
886         1  
887         1  
888         0  
889         1  
890         1  

[891 rows x 3 columns]

¿COMO DETERMINAMOS LOS COMENTARIOS POSITIVOS DE LOS NEGATIVOS EN EL DF?

In [2]:
# Eliminamos la columna irrelevante
df = df.drop(columns=["package_name"])
# Limpiar texto
df['review'] = df['review'].str.strip().str.lower()
# Eliminar nulos por si acaso
df.dropna(inplace=True)

In [3]:
from sklearn.model_selection import train_test_split #declaramos x y para poder entrenar y probar los datos
X = df['review']
y = df['polarity'] #polarity es la columna que determina si el comentario es bueno o malo
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [4]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
vec_model = CountVectorizer()
X_train_vec = vec_model.fit_transform(X_train) #fit.transform se utiliza SOLO en el entrenamiento
X_test_vec = vec_model.transform(X_test) #transform para la prueba

# Que hace CountVectorizer?

crea una bolsa de palabras y columnas


In [5]:
corpus = ["la red se cae todo el tiempo",
          "no tiene buen sonido",
          "la red no es estable",
          "la red tiene lag",
          'red pésima, lag, red mala',
          'ay red red red',
          'lag, lag, lag, lag, lag!']

In [6]:
from sklearn.feature_extraction.text import CountVectorizer
vec = CountVectorizer() #vectoriza el conteo de los datos y los transforma en valores numericos.
X = vec.fit_transform(corpus)
print(vec.get_feature_names_out())
print(X.toarray())

['ay' 'buen' 'cae' 'el' 'es' 'estable' 'la' 'lag' 'mala' 'no' 'pésima'
 'red' 'se' 'sonido' 'tiempo' 'tiene' 'todo']
[[0 0 1 1 0 0 1 0 0 0 0 1 1 0 1 0 1]
 [0 1 0 0 0 0 0 0 0 1 0 0 0 1 0 1 0]
 [0 0 0 0 1 1 1 0 0 1 0 1 0 0 0 0 0]
 [0 0 0 0 0 0 1 1 0 0 0 1 0 0 0 1 0]
 [0 0 0 0 0 0 0 1 1 0 1 2 0 0 0 0 0]
 [1 0 0 0 0 0 0 0 0 0 0 3 0 0 0 0 0]
 [0 0 0 0 0 0 0 5 0 0 0 0 0 0 0 0 0]]


In [7]:
pd.DataFrame(X.toarray(), columns=vec.get_feature_names_out())
#se transforma en un DF para poder visualizarlo mejor

ay  buen  cae  el  es  estable  la  lag  mala  no  pésima  red  se  sonido  \
0   0     0    1   1   0        0   1    0     0   0       0    1   1       0   
1   0     1    0   0   0        0   0    0     0   1       0    0   0       1   
2   0     0    0   0   1        1   1    0     0   1       0    1   0       0   
3   0     0    0   0   0        0   1    1     0   0       0    1   0       0   
4   0     0    0   0   0        0   0    1     1   0       1    2   0       0   
5   1     0    0   0   0        0   0    0     0   0       0    3   0       0   
6   0     0    0   0   0        0   0    5     0   0       0    0   0       0   

   tiempo  tiene  todo  
0       1      0     1  
1       0      1     0  
2       0      0     0  
3       0      1     0  
4       0      0     0  
5       0      0     0  
6       0      0     0

In [8]:
tfidf = TfidfVectorizer() #Tfidf = Term Frequency/ Inverse Document Frequency (importancia que tiene una palabra dentro de todo o todos el/los corpus)
#entre mas veces aparezca una palabra en un doc o review, mayor importancia, pero si aparece en muchos docs o review carecera de valor.
X = tfidf.fit_transform(corpus)
pd.DataFrame(X.toarray(), columns=vec.get_feature_names_out())

ay      buen      cae       el        es   estable        la  \
0  0.00000  0.000000  0.41542  0.41542  0.000000  0.000000  0.294753   
1  0.00000  0.544082  0.00000  0.00000  0.000000  0.000000  0.000000   
2  0.00000  0.000000  0.00000  0.00000  0.535775  0.535775  0.380148   
3  0.00000  0.000000  0.00000  0.00000  0.000000  0.000000  0.503340   
4  0.00000  0.000000  0.00000  0.00000  0.000000  0.000000  0.000000   
5  0.52554  0.000000  0.00000  0.00000  0.000000  0.000000  0.000000   
6  0.00000  0.000000  0.00000  0.00000  0.000000  0.000000  0.000000   

        lag      mala        no    pésima       red       se    sonido  \
0  0.000000  0.000000  0.000000  0.000000  0.224167  0.41542  0.000000   
1  0.000000  0.000000  0.451635  0.000000  0.000000  0.00000  0.544082   
2  0.000000  0.000000  0.444739  0.000000  0.289112  0.00000  0.000000   
3  0.503340  0.000000  0.000000  0.000000  0.382803  0.00000  0.000000   
4  0.370464  0.522126  0.000000  0.522126  0.563495  0.00000  0.000000   
5  0.000000  0.000000  0.000000  0.000000  0.850769  0.00000  0.000000   
6  1.000000  0.000000  0.000000  0.000000  0.000000  0.00000  0.000000   

    tiempo     tiene     todo  
0  0.41542  0.000000  0.41542  
1  0.00000  0.451635  0.00000  
2  0.00000  0.000000  0.00000  
3  0.00000  0.588863  0.00000  
4  0.00000  0.000000  0.00000  
5  0.00000  0.000000  0.00000  
6  0.00000  0.000000  0.00000

In [9]:
from sklearn.naive_bayes import GaussianNB, MultinomialNB, BernoulliNB
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report

# MultinomialNB #cuantas veces la palabra esta, este metodo cuenta
model_mnb = MultinomialNB()
model_mnb.fit(X_train_vec, y_train)
y_pred_mnb = model_mnb.predict(X_test_vec)
print(classification_report(y_test, y_pred_mnb))

# BernoulliNB #si la palabra esta o no esta dentro del review
model_bnb = BernoulliNB()
model_bnb.fit(X_train_vec, y_train)
y_pred_bnb = model_bnb.predict(X_test_vec)
print(classification_report(y_test, y_pred_bnb))


              precision    recall  f1-score   support

           0       0.85      0.95      0.90       126
           1       0.84      0.58      0.69        53

    accuracy                           0.84       179
   macro avg       0.84      0.77      0.79       179
weighted avg       0.84      0.84      0.83       179

              precision    recall  f1-score   support

           0       0.86      0.94      0.90       126
           1       0.81      0.64      0.72        53

    accuracy                           0.85       179
   macro avg       0.84      0.79      0.81       179
weighted avg       0.85      0.85      0.84       179



CONCLUSION:

El modelo Bernoulli clasifica mejor los malos con el 0 en recall del segundo metodo el cual es = 94%

In [25]:

url = "https://raw.githubusercontent.com/4GeeksAcademy/NLP-project-tutorial/main/url_spam.csv"
df = pd.read_csv(url)
df

url  is_spam
0     https://briefingday.us8.list-manage.com/unsubs...     True
1                                https://www.hvper.com/     True
2                    https://briefingday.com/m/v4n3i4f3     True
3      https://briefingday.com/n/20200618/m#commentform    False
4                           https://briefingday.com/fan     True
...                                                 ...      ...
2994  https://www.smartcitiesworld.net/news/news/dee...    False
2995                      https://www.youtube.com/watch     True
2996  https://techcrunch.com/2019/07/04/an-optimisti...    False
2997  https://www.technologyreview.com/2019/12/20/13...    False
2998       https://www.bbc.com/news/technology-51018758    False

[2999 rows x 2 columns]

In [14]:
print(df['is_spam'].value_counts())

is_spam
False    2303
True      696
Name: count, dtype: int64


In [15]:
from sklearn.feature_extraction.text import CountVectorizer
# Vectorizador basado en n-gramas de caracteres
vectorizer = CountVectorizer(analyzer='char_wb', ngram_range=(3,5))  # n-gramas de 3 a 5 caracteres
X = vectorizer.fit_transform(df['url'])
y = df['is_spam']

# analyzer?


In [16]:
from sklearn.feature_extraction.text import CountVectorizer

urls = [
    "https://www.google.com",      # clase 0 (no spam)
    "http://free-money.biz"        # clase 1 (spam)
]

vectorizer = CountVectorizer(analyzer='char', ngram_range=(1, 4))
X = vectorizer.fit_transform(urls)

# el n-grama agrupa cadenas de 1 a 4 caracteres para encontrar patrones

print(vectorizer.get_feature_names_out())

['-' '-m' '-mo' '-mon' '.' '.b' '.bi' '.biz' '.c' '.co' '.com' '.g' '.go'
 '.goo' '/' '//' '//f' '//fr' '//w' '//ww' '/f' '/fr' '/fre' '/w' '/ww'
 '/www' ':' ':/' '://' '://f' '://w' 'b' 'bi' 'biz' 'c' 'co' 'com' 'e'
 'e-' 'e-m' 'e-mo' 'e.' 'e.c' 'e.co' 'ee' 'ee-' 'ee-m' 'ey' 'ey.' 'ey.b'
 'f' 'fr' 'fre' 'free' 'g' 'gl' 'gle' 'gle.' 'go' 'goo' 'goog' 'h' 'ht'
 'htt' 'http' 'i' 'iz' 'l' 'le' 'le.' 'le.c' 'm' 'mo' 'mon' 'mone' 'n'
 'ne' 'ney' 'ney.' 'o' 'og' 'ogl' 'ogle' 'om' 'on' 'one' 'oney' 'oo' 'oog'
 'oogl' 'p' 'p:' 'p:/' 'p://' 'ps' 'ps:' 'ps:/' 'r' 're' 'ree' 'ree-' 's'
 's:' 's:/' 's://' 't' 'tp' 'tp:' 'tp:/' 'tps' 'tps:' 'tt' 'ttp' 'ttp:'
 'ttps' 'w' 'w.' 'w.g' 'w.go' 'ww' 'ww.' 'ww.g' 'www' 'www.' 'y' 'y.'
 'y.b' 'y.bi' 'z']


In [17]:
from sklearn.feature_extraction.text import CountVectorizer

urls = [
    "https://www.google.com",
    "http://free-money.biz"
]
vectorizer = CountVectorizer(analyzer='char_wb', ngram_range=(4, 4))
X = vectorizer.fit_transform(urls)
print(vectorizer.get_feature_names_out())   # no cruza palabras y respeta espacios

[' htt' '-mon' '.biz' '.com' '.goo' '//fr' '//ww' '/fre' '/www' '://f'
 '://w' 'biz ' 'com ' 'e-mo' 'e.co' 'ee-m' 'ey.b' 'free' 'gle.' 'goog'
 'http' 'le.c' 'mone' 'ney.' 'ogle' 'oney' 'oogl' 'p://' 'ps:/' 'ree-'
 's://' 'tp:/' 'tps:' 'ttp:' 'ttps' 'w.go' 'ww.g' 'www.' 'y.bi']


In [18]:
df
X = df['url']
y = df['is_spam']

In [19]:
X_train, X_test, y_train, y_test  = train_test_split(X, y, test_size=0.2, random_state=123)

In [20]:
vectorizer = CountVectorizer(analyzer='char_wb', ngram_range=(4, 4))

In [21]:
X_train = vectorizer.fit_transform(X_train)
X_test = vectorizer.transform(X_test)

In [22]:
from sklearn.naive_bayes import MultinomialNB
model = MultinomialNB()
model.fit(X_train, y_train)

MultinomialNB()

In [23]:
from sklearn.metrics import classification_report, confusion_matrix
y_pred = model.predict(X_test)
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[433  15]
 [ 17 135]]
              precision    recall  f1-score   support

       False       0.96      0.97      0.96       448
        True       0.90      0.89      0.89       152

    accuracy                           0.95       600
   macro avg       0.93      0.93      0.93       600
weighted avg       0.95      0.95      0.95       600



 ### CONCLUSION:
 - El modelo tiene una precisión general del 95%, lo que indica que clasifica correctamente el 95% de las URLs en el conjunto de prueba.

- De todas las URLs que realmente no son spam, el modelo las identifica correctamente el 97% de las veces. Esto es muy bueno, ya que minimiza la posibilidad de bloquear una URL legítima.

- De todas las URLs que son realmente spam, el modelo las identifica correctamente el 89% de las veces. Esto indica que es bastante efectivo en la detección de spam.